File: kotlin/src/main/kotlin/aether/transport/PerTransportMetrics.kt
// SPDX-License-Identifier: MIT

package aether.transport

import java.util.concurrent.atomic.AtomicLong
import kotlin.concurrent.withLock
import kotlin.math.max

/**
 * Real-time per-transport EWMA metrics for adaptive transport selection.
 *
 * α = 0.2: the most-recent sample contributes 20 %; older history decays
 * by a factor of 0.8 per observation.
 *
 * Initial priors:
 *  - ewmaRttMs      = 200 ms   (conservative assumption for unknown links)
 *  - ewmaLossRate   = 0.05     (5 % initial loss assumption)
 *  - ewmaThroughput = 0 bps    (bootstrapped on first successful send)
 *
 * Thread-safe via [java.util.concurrent.locks.ReentrantLock].
 */
class PerTransportMetrics {

    companion object {
        private const val ALPHA = 0.2
    }

    private val lock = java.util.concurrent.locks.ReentrantLock()
    private val _sampleCount = AtomicLong(0)

    private var _ewmaRttMs: Double = 200.0
    private var _ewmaLossRate: Double = 0.05
    private var _ewmaThroughputBps: Double = 0.0

    // ── Read-only accessors ───────────────────────────────────────────────────

    val sampleCount: Long get() = _sampleCount.get()

    val ewmaRttMs: Double
        get() = lock.withLock { _ewmaRttMs }

    val ewmaLossRate: Double
        get() = lock.withLock { _ewmaLossRate }

    val ewmaThroughputBps: Double
        get() = lock.withLock { _ewmaThroughputBps }

    // ── Mutation ──────────────────────────────────────────────────────────────

    /**
     * Update EWMA state from one send observation.
     *
     * @param rttMs            Measured round-trip time in ms (0 = one-way send).
     * @param success          Whether the peer acknowledged receipt.
     * @param bytesTransferred Payload bytes on wire; used for throughput estimate.
     */
    fun recordSample(rttMs: Long, success: Boolean, bytesTransferred: Long) {
        lock.withLock {
            _sampleCount.incrementAndGet()

            if (rttMs > 0) {
                _ewmaRttMs = ALPHA * rttMs + (1 - ALPHA) * _ewmaRttMs
            }

            val lossObs = if (success) 0.0 else 1.0
            _ewmaLossRate = ALPHA * lossObs + (1 - ALPHA) * _ewmaLossRate

            if (success && rttMs > 0L && bytesTransferred > 0L) {
                val tputBps = bytesTransferred * 8.0 * 1000.0 / rttMs
                _ewmaThroughputBps = if (_ewmaThroughputBps < 1.0) {
                    tputBps // bootstrap first sample
                } else {
                    ALPHA * tputBps + (1 - ALPHA) * _ewmaThroughputBps
                }
            }
        }
    }

    // ── Scoring ───────────────────────────────────────────────────────────────

    /**
     * Composite score for transport ranking (higher = better).
     *
     * ```
     * score = (effectiveBps / powerCost) × (1 − lossRate) / max(rttMs, 1)
     * ```
     *
     * `effectiveBps` = max(ewmaThroughputBps, maxBandwidthBps × 0.1)
     * so that zero-sample transports still rank by their declared capacity.
     */
    fun compositeScore(maxBandwidthBps: Long, powerCostRelative: Int): Double {
        val power = max(powerCostRelative, 1)
        val (rtt, loss, tput) = lock.withLock {
            Triple(
                max(_ewmaRttMs, 1.0),
                _ewmaLossRate,
                _ewmaThroughputBps
            )
        }
        val effectiveBps = maxOf(tput, maxBandwidthBps * 0.1)
        return (effectiveBps / power) * (1.0 - loss) / rtt
    }
}

File: python/aether/transport/per_transport_metrics.py
# SPDX-License-Identifier: MIT

"""Real-time per-transport EWMA metrics for adaptive transport selection."""

import threading
from dataclasses import dataclass, field


_ALPHA = 0.2  # EWMA smoothing factor


class PerTransportMetrics:
    """
    Thread-safe, exponentially-weighted moving-average metrics for one transport.

    α = 0.2: the most-recent sample contributes 20 %; older history decays
    by a factor of 0.8 per observation.

    Initial priors:
      - ewma_rtt_ms   = 200 ms   (conservative assumption for unknown links)
      - ewma_loss_rate = 0.05    (5 % initial loss assumption)
      - ewma_throughput_bps = 0  (bootstrapped on first successful send)
    """

    def __init__(self) -> None:
        self._lock = threading.Lock()
        self._sample_count: int = 0
        self._ewma_rtt_ms: float = 200.0
        self._ewma_loss_rate: float = 0.05
        self._ewma_throughput_bps: float = 0.0

    # ── Read-only accessors ───────────────────────────────────────────────────

    @property
    def sample_count(self) -> int:
        """Total samples recorded since this transport started."""
        with self._lock:
            return self._sample_count

    @property
    def ewma_rtt_ms(self) -> float:
        """EWMA round-trip time in milliseconds (lower = better)."""
        with self._lock:
            return self._ewma_rtt_ms

    @property
    def ewma_loss_rate(self) -> float:
        """EWMA packet-loss rate in [0, 1] (lower = better)."""
        with self._lock:
            return self._ewma_loss_rate

    @property
    def ewma_throughput_bps(self) -> float:
        """EWMA throughput in bits per second (higher = better)."""
        with self._lock:
            return self._ewma_throughput_bps

    # ── Mutation ──────────────────────────────────────────────────────────────

    def record_sample(
        self,
        rtt_ms: float,
        success: bool,
        bytes_transferred: int,
    ) -> None:
        """
        Update EWMA state from one send observation.

        Args:
            rtt_ms:            Measured round-trip time in ms (0 = one-way send).
            success:           Whether the peer acknowledged receipt.
            bytes_transferred: Payload bytes on wire; used for throughput.
        """
        with self._lock:
            self._sample_count += 1

            if rtt_ms > 0:
                self._ewma_rtt_ms = (
                    _ALPHA * rtt_ms + (1 - _ALPHA) * self._ewma_rtt_ms
                )

            loss_obs = 0.0 if success else 1.0
            self._ewma_loss_rate = (
                _ALPHA * loss_obs + (1 - _ALPHA) * self._ewma_loss_rate
            )

            if success and rtt_ms > 0 and bytes_transferred > 0:
                tput_bps = bytes_transferred * 8.0 * 1000.0 / rtt_ms
                if self._ewma_throughput_bps < 1.0:
                    self._ewma_throughput_bps = tput_bps  # bootstrap
                else:
                    self._ewma_throughput_bps = (
                        _ALPHA * tput_bps
                        + (1 - _ALPHA) * self._ewma_throughput_bps
                    )

    # ── Scoring ───────────────────────────────────────────────────────────────

    def composite_score(
        self,
        max_bandwidth_bps: int,
        power_cost_relative: int,
    ) -> float:
        """
        Composite score (higher = better transport to select right now).

        Formula:
            score = (effective_bps / power_cost) × (1 − loss_rate) / max(rtt_ms, 1)

        where effective_bps = max(ewma_throughput_bps, max_bandwidth_bps × 0.1)
        so zero-sample transports still rank by their declared capacity.
        """
        power = max(power_cost_relative, 1)
        with self._lock:
            rtt = max(self._ewma_rtt_ms, 1.0)
            loss = self._ewma_loss_rate
            tput = self._ewma_throughput_bps

        effective_bps = max(tput, max_bandwidth_bps * 0.1)
        return (effective_bps / power) * (1.0 - loss) / rtt
